### PR TITLE
Use `go list -m -versions` to determine available versions of a go module

### PR DESCRIPTION
**Pull request by mctofu**
This removes usage of the [native helper](https://github.com/dependabot/dependabot-core/blob/417dd1de9ac6ee9f8e366031a8efb288f6484291/go_modules/helpers/updatechecker/main.go) to resolve available go module versions. Instead we'll use the existing go tooling which has added [support for this](https://golang.org/ref/mod#go-list-m). This has a few benefits:
- No longer need to maintain https://github.com/dependabot/gomodules-extracted or make calls to internal go libs
- No longer need to manually handle exclude directives
- Retract directives are now supported so Dependabot won't propose updates to retracted versions

We previously attempted this change in https://github.com/dependabot/dependabot-core/pull/3630 but the approach was abandoned after we found the change was negatively impacting performance. It was [further noted](https://github.com/dependabot/dependabot-core/pull/3630#issuecomment-832678444) that using `go list -m -versions` was causing the go tooling to download all of the modules referenced in `go.mod` and this accounted for the bulk of the performance hit. This appears to only happen when `GOPRIVATE=*` is set which we currently need to do to support private modules.

As a workaround I found that providing a `go.mod` with no requirements was a way to sidestep the performance issue. We still end up getting the version information we are looking for and no extra modules are downloaded! The performance is close to the native version as well.

### Example output / performance
<details>
  <summary>Using native helper</summary>
  
```
[dependabot-core-dev] /tmp/go $ rm -rf /home/dependabot/go/pkg/mod/cache
[dependabot-core-dev] /tmp/go $ cat go.mod
module foo

require github.com/dependabot-fixtures/go-modules-lib v1.0.0

go 1.15
[dependabot-core-dev] /tmp/go $ time echo '{"function":"getVersions","args":{"dependency":{"name":"github.com/dependabot-fixtures/go-modules-lib","version":"v1.0.0"}}}'  | GOPRIVATE='*' /opt/go_modules/bin/helper
{"result":["v1.0.0","v1.0.1","v1.0.5","v1.0.6","v1.1.0","v1.2.0-pre1","v1.2.0-pre2"]}
real    0m1.017s
user    0m0.105s
sys     0m0.087s
[dependabot-core-dev] /tmp/go $ du -sk /home/dependabot/go/pkg/mod/cache
176     /home/dependabot/go/pkg/mod/cache
```

</details>

<details>
  <summary>Using `go list -m -versions` with project `go.mod`</summary>
  
```
[dependabot-core-dev] /tmp/go $ rm -rf /home/dependabot/go/pkg/mod/cache
[dependabot-core-dev] /tmp/go $ cat go.mod
module foo

require github.com/dependabot-fixtures/go-modules-lib v1.0.0

go 1.15
[dependabot-core-dev] /tmp/go $ time GOPRIVATE='*' go list -m -versions -json github.com/dependabot-fixtures/go-modules-lib
{
        "Path": "github.com/dependabot-fixtures/go-modules-lib",
        "Version": "v1.0.0",
        "Versions": [
                "v1.0.0",
                "v1.0.1",
                "v1.0.5",
                "v1.0.6",
                "v1.1.0",
                "v1.2.0-pre1",
                "v1.2.0-pre2"
        ],
        "Time": "2018-10-18T21:48:48Z",
        "GoMod": "/home/dependabot/go/pkg/mod/cache/download/github.com/dependabot-fixtures/go-modules-lib/@v/v1.0.0.mod"
}

real    0m8.828s
user    0m1.882s
sys     0m1.465s
[dependabot-core-dev] /tmp/go $ du -sk /home/dependabot/go/pkg/mod/cache
24844   /home/dependabot/go/pkg/mod/cache
```

</details>

<details>
  <summary>Using `go list -m -versions` with stub `go.mod` (this PR)</summary>
  
```
[dependabot-core-dev] /tmp/go $ rm -rf /home/dependabot/go/pkg/mod/cache
[dependabot-core-dev] /tmp/go $ cat go.mod
module foo
[dependabot-core-dev] /tmp/go $ time GOPRIVATE='*' go list -m -versions -json github.com/dependabot-fixtures/go-modules-lib
{
        "Path": "github.com/dependabot-fixtures/go-modules-lib",
        "Versions": [
                "v1.0.0",
                "v1.0.1",
                "v1.0.5",
                "v1.0.6",
                "v1.1.0",
                "v1.2.0-pre1",
                "v1.2.0-pre2"
        ]
}

real    0m1.705s
user    0m0.140s
sys     0m0.135s
[dependabot-core-dev] /tmp/go $ du -sk /home/dependabot/go/pkg/mod/cache
228     /home/dependabot/go/pkg/mod/cache
```

</details>

### Other notes
- This change revealed that the [v2](https://github.com/dependabot-fixtures/go-modules-lib/blob/v2.0.0/go.mod) release of our test fixture wasn't actually valid. I created a valid [v3](https://github.com/dependabot-fixtures/go-modules-lib/blob/v3.0.0/go.mod) release and added handling for invalid version errors.
- I believe the usage of `SharedHelpers.run_shell_command` will escape the args being passed to `go mod list -m -versions` and `go mod edit` to prevent command injection. I'd appreciate a 2nd look at this though.

### CODE REVIEW

**Comment by jeffwidman** [comment]
> As a workaround I found that providing a go.mod with no requirements was a way to sidestep the performance issue. We still end up getting the version information we are looking for and no extra modules are downloaded!

Very clever!

**Review by jurre — APPROVED** [review]
This looks great, so glad we're able to get rid of `gomodules-extracted`!

> I believe the usage of SharedHelpers.run_shell_command will escape the args being passed to go mod list -m -versions and go mod edit to prevent command injection. I'd appreciate a 2nd look at this though.

That's right, unless we explicitly tell `run_shell_command` to `allow_unsafe_shell_command` it runs any `command` we pass through `Shellwords` which should protect against any command injection 👍

**Comment by weitzj** [comment]
Looking forward to this release.

File: common/lib/dependabot/shared_helpers.rb [hunk 1]
```diff
@@ -280,10 +280,11 @@ def self.reset_global_git_config(backup_path)
       FileUtils.mv(backup_path, GIT_CONFIG_GLOBAL_PATH)
     end
 
-    def self.run_shell_command(command, allow_unsafe_shell_command: false)
+    def self.run_shell_command(command, allow_unsafe_shell_command: false, env: nil)
       start = Time.now
       cmd = allow_unsafe_shell_command ? command : escape_command(command)
-      stdout, process = Open3.capture2e(cmd)
+      env_cmd = [env, cmd].compact
+      stdout, process = Open3.capture2e(*env_cmd)
```
Comment (mattt) [inline common/lib/dependabot/shared_helpers.rb:287]: Could we rephrase this to make the intent of this a bit clearer? It took me a few minutes to understand what was happening here.

At first, I thought _"`env` is something like `TEST_ENV=prefix`, so this concatenates the original command"_. But then I realized that `env` is a hash, and had to consult the [docs for `Process#spawn`](https://docs.ruby-lang.org/en/2.0.0/Process.html#method-c-spawn) to understand how `compact` is called here to mitigate the awkwardness of `#spawn` taking `env` as an optional first argument.

Would something like this work instead?

```ruby
def self.run_shell_command(command, allow_unsafe_shell_command: false, env: {})
      start = Time.now
      cmd = allow_unsafe_shell_command ? command : escape_command(command)
      stdout, process = Open3.capture2e(env, cmd)
```

Comment (landongrindheim) [reply common/lib/dependabot/shared_helpers.rb:287]: I think using a default of `env: {}` in the method signature might clear this up as well. We don't have a guarantee against `nil` being passed in, but we catch the cases where it is not supplied 🙂

Comment (jurre) [reply common/lib/dependabot/shared_helpers.rb:287]: 👍 the optional first argument has thrown me off a few times as well, it's such a weird API

Comment (mctofu) [reply common/lib/dependabot/shared_helpers.rb]: Love these ideas! Updated in https://github.com/dependabot/dependabot-core/pull/4434/commits/16395d30cedb5554090873ece20d99321cea4fc7.

There is another usage of compact at https://github.com/dependabot/dependabot-core/blob/ebe338ed2bd1d09b2c86510f9e44cc0cf642e18b/common/lib/dependabot/shared_helpers.rb#L93-L94 but I'm going to leave that for a future PR.

**Review by mattt — APPROVED** [review]

File: go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb [hunk 2]
```diff
@@ -73,23 +75,22 @@ def fetch_lowest_security_fix_version
         def available_versions
           SharedHelpers.in_a_temporary_directory do
             SharedHelpers.with_git_configured(credentials: credentials) do
-              File.write("go.mod", go_mod.content)
+              manifest = parse_manifest
+
+              # Set up an empty go.mod so 'go list -m' won't attempt to download dependencies. This
```
Comment (mctofu) [inline go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb:80]: One other observation from testing is that we can also run `go list -m -versions` without a `go.mod` and it works the same way. However, it won't know about the exclude directives and won't be able to apply them for us so creating a `go.mod` with the excludes is still preferred.

File: go_modules/helpers/main.go [hunk 3]
```diff
@@ -32,10 +31,6 @@ func main() {
 		funcErr error
 	)
 	switch helperParams.Function {
-	case "getVersions":
-		var args updatechecker.Args
-		parseArgs(helperParams.Args, &args)
-		funcOut, funcErr = updatechecker.GetVersions(&args)
 	case "getVcsRemoteForImport":
 		var args importresolver.Args
 		parseArgs(helperParams.Args, &args)
```
Comment (jeffwidman) [inline go_modules/helpers/main.go:36]: May be able to delete this as well, I filed https://github.com/dependabot/dependabot-core/issues/4448 for that

**Comment by jeffwidman** [comment]
Nicely done! 

I opened https://github.com/dependabot/gomodules-extracted/pull/17 and https://github.com/dependabot/gomodules-extracted/issues/18 as followups to this.

**Comment by mctofu** [comment]
I noticed one unanticipated side effect where Dependabot now updates major versions for `+incompatible` versions: https://github.com/dependabot/dependabot-core/issues/4453
